Source/Core/DolphinQt/FIFO/FIFOAnalyzer.cpp

```diff
@@ -158,155 +164,251 @@ void FIFOAnalyzer::UpdateTree()
 
-      case OpcodeDecoder::GX_LOAD_XF_REG:
-      {
-        u32 cmd2 = Common::swap32(objectdata);
-        objectdata += 4;
+    case OpcodeDecoder::GX_LOAD_BP_REG:
+    {
+      const u8 cmd2 = object[object_offset++];
+      const u32 cmddata = Common::swap24(&object[object_offset]);
+      object_offset += 3;
 
-        u8 streamSize = ((cmd2 >> 16) & 15) + 1;
+      const auto [name, desc] = GetBPRegInfo(cmd2, cmddata);
 
-        const u8* stream_start = objectdata;
-        const u8* stream_end = stream_start + streamSize * 4;
+      new_label = QStringLiteral("BP  %1 %2   %3")
+                      .arg(cmd2, 2, 16, QLatin1Char('0'))
+                      .arg(cmddata, 6, 16, QLatin1Char('0'))
+                      .arg(QString::fromStdString(name));
+    }
+    break;
 
-        new_label = QStringLiteral("XF  %1  ").arg(cmd2, 8, 16, QLatin1Char('0'));
-        while (objectdata < stream_end)
-        {
-          new_label += QStringLiteral("%1").arg(*objectdata++, 2, 16, QLatin1Char('0'));
+    default:
+      if ((command & 0xC0) == 0x80)
+      {
+        // Object primitive data
 
-          if (((objectdata - stream_start) % 4) == 0)
-            new_label += QLatin1Char(' ');
+        // object_nonprim_size is found by FIFOPlaybackAnalyzer, and we need it to be correct
+        // because we use the object's end (also calculated by FIFOPlaybackAnalyzer) to determine
+        // the length, to avoid needing to know the vertex format here.
+        if (start_offset != object_nonprim_size)
+        {
+          PanicAlertFmt("The primitive command should be the last command in the object, at "
+                        "{:08x}, but it instead was at {:08x}",
+                        object_nonprim_size, start_offset);
         }
-      }
-      break;
 
-      case OpcodeDecoder::GX_LOAD_INDX_A:
-      case OpcodeDecoder::GX_LOAD_INDX_B:
-      case OpcodeDecoder::GX_LOAD_INDX_C:
-      case OpcodeDecoder::GX_LOAD_INDX_D:
-      {
-        objectdata += 4;
-        new_label = (command == OpcodeDecoder::GX_LOAD_INDX_A) ?
-                        QStringLiteral("LOAD INDX A") :
-                        (command == OpcodeDecoder::GX_LOAD_INDX_B) ?
-                        QStringLiteral("LOAD INDX B") :
-                        (command == OpcodeDecoder::GX_LOAD_INDX_C) ? QStringLiteral("LOAD INDX C") :
-                                                                     QStringLiteral("LOAD INDX D");
-      }
-      break;
+        const auto name = GetPrimitiveName(command);
 
-      case OpcodeDecoder::GX_CMD_CALL_DL:
-        // The recorder should have expanded display lists into the fifo stream and skipped the
-        // call to start them
-        // That is done to make it easier to track where memory is updated
-        ASSERT(false);
-        objectdata += 8;
-        new_label = QStringLiteral("CALL DL");
-        break;
+        const u16 vertex_count = Common::swap16(fifo_frame.fifoData[object_offset]);
```

Comment:
This is a typo, for two reasons: one, it should use `object[object_offset]`, but two, it's passing a u8 to `swap16` when it should be passing a `u8*`.  I've made this mistake several times and it is stupidly easy to miss unfortunately.  This one caused the number of vertices to usually be zero, which really confused me.